File: least-steep-path-finder/pathfinder.py
# live version: https://repl.it/repls/TerrificIckyCode

import os
import time
import random

def clear_screen():
    if os.name == 'nt':
        os.system('cls')
    else:
        os.system('clear')

clear_screen()
grid_size = int(input('Enter Grid Size: '))
grid = []

# x = 0
# for i in range(grid_size):
#     grid.append([])
#     for j in range(grid_size):
#         grid[i].append(x)
#         x += 1

for i in range(grid_size):
    grid.append([])
    for j in range(grid_size):
        grid[i].append(random.randint(1, 99))

# grid_size = 8
# grid = [
#     [25, 92, 47, 61, 13, 11, 25, 68],
#     [88, 99, 61, 20, 88, 99, 61, 20],
#     [13, 11, 25, 68, 12, 34, 37, 40],
#     [12, 34, 37, 40, 25, 92, 47, 61],
#     [88, 79, 61, 20, 88, 99, 61, 20],
#     [13, 11, 25, 68, 12, 34, 37, 40],
#     [25, 92, 61, 61, 13, 11, 25, 68],
#     [12, 34, 37, 40, 25, 92, 47, 61],
# ]

clear_screen()
print()
print('GRID')
print()
print("\t", end="")
for i in range(grid_size):
    print(str(i) + "\t", end="")
print()
print("\t", end="")
for i in range(grid_size):
    print("_\t", end="")
for i, v in enumerate(grid):
    print()
    print()
    print(str(i) + "|\t", end="")
    for j, w in enumerate(v):
        print(str(grid[i][j]) + '\t', end="")
print()
print()
print()

print('Starting Point')
input_row = int(input("Enter Row Number: "))
input_col = int(input("Enter Col Number: "))
start = (input_row, input_col, grid[input_row][input_col])
print()
print('Destination Point')
input_row = int(input("Enter Row Number: "))
input_col = int(input("Enter Col Number: "))
destination = (input_row, input_col, grid[input_row][input_col])
current = start
prev_steps = [start]
taken_steps = [start]
iterations = 1
num_of_steps = 1

for row in grid:
    print(row)
while (current[0] != destination[0] 
        or current[1] != destination[1] 
        or current[2] != destination[2]):
    row, column = current[0], current[1]
    directions = {
        'up': (row - 1, column) if row - 1 < grid_size and row - 1 >= 0 else None,
        'down': (row + 1, column) if row + 1 < grid_size and row + 1 >= 0 else None,
        'left': (row, column - 1) if column - 1 < grid_size and column - 1 >= 0 else None,
        'right': (row, column + 1) if column + 1 < grid_size and column + 1 >= 0 else None,
    }
    for direction in directions:
        if directions[direction] is not None:
            directions[direction] = (
                directions[direction][0], 
                directions[direction][1], 
                grid[directions[direction][0]][directions[direction][1]],
            )
            
            if directions[direction] in prev_steps:
                directions[direction] = None
            
    if directions['up'] is None\
        and directions['down'] is None\
        and directions['left'] is None\
        and directions['right'] is None\
        and len(prev_steps) > 2:
            current = start
            prev_steps = prev_steps[
                -iterations if iterations < len(prev_steps) else -1 * random.randint(0, len(prev_steps))
                ::]
            # prev_steps = prev_steps[-iterations::]
            taken_steps = []
            num_of_steps = 0
            iterations += 1
            
            clear_screen()
            print()
            print('GRID')
            print()
            print("\t", end="")
            for i in range(grid_size):
                print(str(i) + "\t", end="")
            print()
            print("\t", end="")
            for i in range(grid_size):
                print("_\t", end="")
            for i, v in enumerate(grid):
                print()
                print()
                print(str(i) + "|\t", end="")
                for j, w in enumerate(v):
                    temp_tuple = (i, j, w)
                    if temp_tuple in taken_steps:
                        print(str(grid[i][j]) + "^\t", end="")
                    else:
                        if temp_tuple in prev_steps:
                            print("_\t", end="")
                        else:
                            print(str(grid[i][j]) + '\t', end="")
            print()
            print()
            print()
            print('Trapped! If this does not go away in 2 seconds, then Press Ctrl+C to Quit!')
            time.sleep(0.5)
            
            # continue

    next_step = False
    pref_steps = []

    if current[2] < destination[2]:
        for direction in directions:
            if directions[direction] is not None and directions[direction][2] >= current[2]:
                pref_steps.append(directions[direction])
        
        if len(pref_steps) > 1:
            for direction in pref_steps:
                for direction_ in pref_steps:
                    if direction_[2] <= direction[2]:
                        next_step = direction_
        elif len(pref_steps) == 1:
            next_step = pref_steps[0]
        else:
            pref_steps = []
            for direction in directions:
                if directions[direction] is not None and directions[direction][2] <= current[2]:
                    pref_steps.append(directions[direction])
            if len(pref_steps) > 1:
                for direction in pref_steps:
                    for direction_ in pref_steps:
                        if direction_[2] >= direction[2]:
                            next_step = direction_
            elif len(pref_steps) == 1:
                next_step = pref_steps[0]
        
        if next_step is not False:
            current = next_step

    elif current[2] >= destination[2]:
        for direction in directions:
            if directions[direction] is not None and directions[direction][2] <= current[2]:
                pref_steps.append(directions[direction])

        if len(pref_steps) > 1:
            for direction in pref_steps:
                for direction_ in pref_steps:
                    if direction_[2] >= direction[2]:
                        next_step = direction_
        elif len(pref_steps) == 1:
            next_step = pref_steps[0]
        else:
            pref_steps = []
            for direction in directions:
                if directions[direction] is not None and directions[direction][2] >= current[2]:
                    pref_steps.append(directions[direction])
            if len(pref_steps) > 1:
                for direction in pref_steps:
                    for direction_ in pref_steps:
                        if direction_[2] <= direction[2]:
                            next_step = direction_
            elif len(pref_steps) == 1:
                next_step = pref_steps[0]
            
        if next_step is not False:
            current = next_step

    taken_steps.append(current)
    if current not in prev_steps:
        prev_steps.append(current)
    
    clear_screen()
    print()
    print('GRID')
    print()
    print("\t", end="")
    for i in range(grid_size):
        print(str(i) + "\t", end="")
    print()
    print("\t", end="")
    for i in range(grid_size):
        print("_\t", end="")
    for i, v in enumerate(grid):
        print()
        print()
        print(str(i) + "|\t", end="")
        for j, w in enumerate(v):
            temp_tuple = (i, j, w)
            if temp_tuple in taken_steps:
                print(str(grid[i][j]) + "^\t", end="")
            else:
                if temp_tuple in prev_steps:
                    print("_\t", end="")
                else:
                    print(str(grid[i][j]) + '\t', end="")
    print()
    print()
    num_of_steps += 1
    time.sleep(0.2)

clear_screen()
print()
print('FINAL GRID')
print()
print("\t", end="")
for i in range(grid_size):
    print(str(i) + "\t", end="")
print()
print("\t", end="")
for i in range(grid_size):
    print("_\t", end="")
for i, v in enumerate(grid):
    print()
    print()
    print(str(i) + "|\t", end="")
    for j, w in enumerate(v):
        temp_tuple = (i, j, w)
        if temp_tuple in taken_steps:
            print(str(grid[i][j]) + "^" + str(taken_steps.index(temp_tuple)) + "\t", end="")
        else:
            print(str(grid[i][j]) + '\t', end="")
print()
print()
print()
print('Taken Steps: ', taken_steps)
print('Iterations: ' + str(iterations))
print('Num of steps: ', num_of_steps)

File: least-steep-path-finder/pathfinder.js
const gridSizeInput = document.querySelector('#gridSizeInput');
const gridGenBtn = document.querySelector('#gridGenBtn');
const gridCtr = document.querySelector('#gridCtr');

let gridSize = 0;
let startPoint = null;
let destPoint = null;
let curPoint = null;
let grid = [];
let prevSteps = []
let takenSteps = []
let iterations = 1
let numOfSteps = 1

let isStartSelected = false;
let isDestSelected = false;

gridGenBtn.addEventListener('click', (ev) => {
    gridSize = parseInt(gridSizeInput.value, 10);
    gridSize = gridSize < 4 ? 4 : gridSize;
    gridCtr.innerHTML = '';
    grid = [];
    for (let i = 0; i < gridSize; i++) {
        grid.push([]);
        for (let j = 0; j < gridSize; j++) {
            const randNum = getRandomInt(100);
            grid[i].push(randNum);
            const gridBox = `
                <div 
                    style="width: 30px; height: 30px; display: inline-block; padding: 2px; background: #efefef; color: black; border: 1px solid black; font-size: 14px; text-align: center; user-select: none;"
                    id="${'gridBox_' + i.toString() + '_' + j.toString()}"
                    data-row="${i}"
                    data-col="${j}"
                    data-val="${randNum}"
                ><strong style="float: right; font-size: 8px;"></strong>
                <span style="display: inline; vertical-align: bottom; float: left;">${randNum}</span></div>
            `;
            gridCtr.innerHTML += gridBox;
        }
        gridCtr.innerHTML += '<br />';
    }
    const gridBoxes = gridCtr.querySelectorAll('div');
    gridBoxes.forEach(elem => { 
        elem.addEventListener('click', (e) => {
            e.stopPropagation();
            if (!isStartSelected) {
                startPoint = [
                    parseInt(elem.getAttribute('data-row'), 10),
                    parseInt(elem.getAttribute('data-col'), 10),
                    parseInt(elem.getAttribute('data-val'), 10),
                ];
                elem.style.background = 'green';
                elem.style.color = 'white';
                isStartSelected = true;
                document.querySelector('#instructions')
                    .innerHTML = `
                        <em><b>Instructions: </b></em><br />Click a box again to pick as Destination Point
                    `;
            } else if (!isDestSelected) {
                destPoint = [
                    parseInt(elem.getAttribute('data-row'), 10),
                    parseInt(elem.getAttribute('data-col'), 10),
                    parseInt(elem.getAttribute('data-val'), 10),
                ];
                elem.style.background = 'red';
                elem.style.color = 'white';
                isDestSelected = true;
                document.querySelector('#instructions')
                    .innerHTML = `
                        <em><b>Instructions: </b></em><br />Click Start Button
                    `;
                document.querySelector('#results #startBtn').style.display = '';
                document.querySelector('#results #startBtn').addEventListener('click', (evt) => {
                    startProcessing();
                    evt.target.setAttribute('disabled', '');
                });
            }
        }, false);
    });
    document.querySelector('#instructions')
        .innerHTML = `
            <em><b>Instructions: </b></em><br />Click a box to pick as Starting Point
        `;
    gridGenBtn.setAttribute('disabled', '');
});


function startProcessing() {
    
    curPoint = startPoint;
    prevSteps = [startPoint]
    takenSteps = [startPoint]
    
    let reRun = setInterval(() => {
        makePath();
    }, 20);

    function makePath() {

        if (curPoint[0] === destPoint[0] 
            && curPoint[1] === destPoint[1] 
            && curPoint[2] === destPoint[2]) {
                clearInterval(reRun);
                document.querySelector('#results').innerHTML = `
                    <br />
                    <a href="index.html">Reload</a> <br /> <br />
                    Taken Steps: ${JSON.stringify(takenSteps)} <br />
                    Number of Steps: ${takenSteps.length} <br />
                    Iterations took to get valid path: ${iterations} <br />
                `;
                document.querySelector('#instructions')
                    .innerHTML = `
                        <em><b>Instructions: </b></em><br />Process Completed! 'Reload' to Start Again
                    `;
                return;
        }

        let row = curPoint[0];
        let column = curPoint[1];

        let directions = {
            'up': row - 1 < gridSize && row - 1 >= 0 ? [row - 1, column] : null,
            'down': row + 1 < gridSize && row + 1 >= 0 ? [row + 1, column] : null,
            'left': column - 1 < gridSize && column - 1 >= 0 ? [row, column - 1] : null,
            'right': column + 1 < gridSize && column + 1 >= 0 ? [row, column + 1] : null,
        };
        for (let direction in directions ) {
            if (directions[direction] !== null) {
                directions[direction] = [
                    directions[direction][0], 
                    directions[direction][1], 
                    grid[directions[direction][0]][directions[direction][1]],
                ];
                
                for (let i = 0; i < prevSteps.length; i++) {
                    if (isEqual(directions[direction], prevSteps[i])) {
                        directions[direction] = null;
                    }
                }
            }
        }

        if (directions['up'] === null
                && directions['down'] === null
                && directions['left'] === null
                && directions['right'] === null
                && prevSteps.length > 2) {
            curPoint = startPoint;
            prevSteps = prevSteps.slice(
                iterations < prevSteps.length ? -iterations : -1 * getRandomInt(prevSteps.length),
                prevSteps.length
            );
            takenSteps = [];
            numOfSteps = 0;
            iterations += 1;

            updateGrid();
        }

        let nextStep = null;
        let prefSteps = [];

        if (curPoint[2] < destPoint[2]) {
            for (let direction in directions) {
                if (directions[direction] !== null && directions[direction][2] >= curPoint[2]) {
                    prefSteps.push(directions[direction]);
                }
            }
            if (prefSteps.length > 1) {
                for (let direction in prefSteps) {
                    for (let direction_ in prefSteps) {
                        direction = parseInt(direction, 10);
                        direction_ = parseInt(direction_, 10);
                        if (prefSteps[direction_][2] < prefSteps[direction][2]) {
                            nextStep = prefSteps[direction_];
                        }
                    }
                }
            } else if (prefSteps.length === 1) {
                nextStep = prefSteps[0];
            } else {
                prefSteps = []
                for (let direction in directions) {
                    if (directions[direction] !== null && directions[direction][2] <= curPoint[2]) {
                        prefSteps.push(directions[direction]);
                    }
                }
                if (prefSteps.length > 1) {
                    for (let direction in prefSteps) {
                        for (let direction_ in prefSteps) {
                            direction = parseInt(direction, 10);
                            direction_ = parseInt(direction_, 10);
                            if (prefSteps[direction_][2] > prefSteps[direction][2]) {
                                nextStep = prefSteps[direction_];
                            }
                        }
                    }
                } else if (prefSteps.length === 1) {
                    nextStep = prefSteps[0];
                }
            }

            if (nextStep !== null) {
                curPoint = nextStep;
            } else {
                for (let direction in directions) {
                    if (directions[direction] !== null) {
                        curPoint = directions[direction];
                        break;
                    }
                }
            }
        } else if (curPoint[2] >= destPoint[2]) {
            for (let direction in directions) {
                if (directions[direction] !== null && directions[direction][2] <= curPoint[2]) {
                    prefSteps.push(directions[direction]);
                }
            }
            if (prefSteps.length > 1) {
                for (let direction in prefSteps) {
                    for (let direction_ in prefSteps) {
                        direction = parseInt(direction, 10);
                        direction_ = parseInt(direction_, 10);
                        if (prefSteps[direction_][2] > prefSteps[direction][2]) {
                            nextStep = prefSteps[direction_];
                        }
                    }
                }
            } else if (prefSteps.length === 1) {
                nextStep = prefSteps[0];
            } else {
                prefSteps = [];
                for (let direction in directions) {
                    if (directions[direction] !== null && directions[direction][2] >= curPoint[2]) {
                        prefSteps.push(directions[direction]);
                    }
                }
                if (prefSteps.length > 1) {
                    for (let direction in prefSteps) {
                        for (let direction_ in prefSteps) {
                            direction = parseInt(direction, 10);
                            direction_ = parseInt(direction_, 10);
                            if (prefSteps[direction_][2] < prefSteps[direction][2]) {
                                nextStep = prefSteps[direction_];
                            }
                        }
                    }
                } else if (prefSteps.length === 1) {
                    nextStep = prefSteps[0];
                }
            }

            if (nextStep !== null) {
                curPoint = nextStep;
            } else {
                for (let direction in directions) {
                    if (directions[direction] !== null) {
                        curPoint = directions[direction];
                        break;
                    }
                }
            }
        }

        takenSteps.push(curPoint);
        let curPointInPrevSteps = true;
        for (let step in prevSteps) {
            step = parseInt(step, 10);
            if (isEqual(curPoint, prevSteps[step])) {
                curPointInPrevSteps = true;
            } else {
                curPointInPrevSteps = false;
            }
        }
        if (!curPointInPrevSteps)
            prevSteps.push(curPoint);
        
        updateGrid();

    }
}

function updateGrid() {
    for (let i in grid) {
        for (let j in grid[i]) {
            i = parseInt(i, 10);
            j = parseInt(j, 10);
            const tempPoint = [i, j, grid[i][j]];
            if (!isEqual(tempPoint, startPoint) && !isEqual(tempPoint, destPoint)) {
                let elem = document.querySelector(`#gridBox_${i}_${j}`);
                elem.style.background = '#efefef';
                elem.style.color = 'black';
                elem.querySelector('strong').innerHTML = ``;
            }
        }
    }
    for (let i in prevSteps) {
        if (!isEqual(prevSteps[i], startPoint) && !isEqual(prevSteps[i], destPoint)) {
            let elem = document.querySelector(`#gridBox_${prevSteps[i][0]}_${prevSteps[i][1]}`);
            elem.style.background = '#aaaaaa';
            elem.style.color = '#efefef';
            elem.querySelector('strong').innerHTML = ``;
        }
    }
    for (let i in takenSteps) {
        let elem = document.querySelector(`#gridBox_${takenSteps[i][0]}_${takenSteps[i][1]}`);
        if (!isEqual(takenSteps[i], startPoint) && !isEqual(takenSteps[i], destPoint)) {
            elem.style.background = 'blue';
            elem.style.color = 'white';
        }
        elem.querySelector('strong').innerHTML = `${i}`;
    }
}


// taken from MDN
function getRandomInt(max) {
    return Math.floor(Math.random() * Math.floor(max));
}


// taken from https://gomakethings.com/check-if-two-arrays-or-objects-are-equal-with-javascript/
var isEqual = function (value, other) {

	// Get the value type
	var type = Object.prototype.toString.call(value);

	// If the two objects are not the same type, return false
	if (type !== Object.prototype.toString.call(other)) return false;

	// If items are not an object or array, return false
	if (['[object Array]', '[object Object]'].indexOf(type) < 0) return false;

	// Compare the length of the length of the two items
	var valueLen = type === '[object Array]' ? value.length : Object.keys(value).length;
	var otherLen = type === '[object Array]' ? other.length : Object.keys(other).length;
	if (valueLen !== otherLen) return false;

	// Compare two items
	var compare = function (item1, item2) {

		// Get the object type
		var itemType = Object.prototype.toString.call(item1);

		// If an object or array, compare recursively
		if (['[object Array]', '[object Object]'].indexOf(itemType) >= 0) {
			if (!isEqual(item1, item2)) return false;
		}

		// Otherwise, do a simple comparison
		else {

			// If the two items are not the same type, return false
			if (itemType !== Object.prototype.toString.call(item2)) return false;

			// Else if it's a function, convert to a string and compare
			// Otherwise, just compare
			if (itemType === '[object Function]') {
				if (item1.toString() !== item2.toString()) return false;
			} else {
				if (item1 !== item2) return false;
			}

		}
	};

	// Compare properties
	if (type === '[object Array]') {
		for (var i = 0; i < valueLen; i++) {
			if (compare(value[i], other[i]) === false) return false;
		}
	} else {
		for (var key in value) {
			if (value.hasOwnProperty(key)) {
				if (compare(value[key], other[key]) === false) return false;
			}
		}
	}

	// If nothing failed, return true
	return true;

};